internal/app/go-filecoin/plumbing/msg/waiter.go

```diff
@@ -148,55 +151,64 @@ func (w *Waiter) waitForMessage(ctx context.Context, ch <-chan interface{}, msgC
 	}
 }
 
-func (w *Waiter) receiptForTipset(ctx context.Context, ts block.TipSet, msgCid cid.Cid) (*ChainMessage, bool, error) {
-	tsMessages := make([][]*types.SignedMessage, ts.Len())
+func (w *Waiter) receiptForTipset(ctx context.Context, ts block.TipSet, targetMsg cid.Cid) (*ChainMessage, bool, error) {
+	// The targetMsg might be the CID of either a signed SECP message or an unsigned
+	// BLS message.
+	// This accumulates the CIDs of the messages as they appear on chain (signed or unsigned)
+	// but then unwraps them all, obtaining the CID of the unwrapped SECP message body if
+	// applicable. This unwrapped message CID is then used to find the target message in the
+	// unwrapped de-duplicated tipset messages, and thence the corresponding receipt by index.
+	tsMessages := make([][]*types.UnsignedMessage, ts.Len())
 	for i := 0; i < ts.Len(); i++ {
 		blk := ts.At(i)
 		secpMsgs, blsMsgs, err := w.messageProvider.LoadMessages(ctx, blk.Messages)
 		if err != nil {
 			return nil, false, err
 		}
 
-		cids := make([]cid.Cid, len(blsMsgs)+len(secpMsgs))
-		blkMsgs := make([]*types.SignedMessage, len(blsMsgs)+len(secpMsgs))
+		originalCids := make([]cid.Cid, len(blsMsgs)+len(secpMsgs))
+		unwrappedMsgs := make([]*types.UnsignedMessage, len(blsMsgs)+len(secpMsgs))
+		wrappedMsgs := make([]*types.SignedMessage, len(blsMsgs)+len(secpMsgs))
```

Comment:
FYI @acruikshank rebasing against your earlier change caused some gymnastics in the waiter, as de-duping now occurs over unsigned messages.

Changing the waiter callback result value from SignedMessage to UnsignedMessage could reduce the pain a bit, but is also slightly wrong.

We might need a union-style type that can represent either signed or unsigned, presenting a canonical CID in either case.